gebrd implicit test

tfalders opened the pull request:
Addresses item 1 of SWDEV-247005, and should address SWDEV-246839.

SWDEV-246839 is affecting QA, so the sooner we can get this change into staging, the better.

cgmb commented:
Not a comment about the code, but maybe I can learn something while this work is fresh in your mind.

1. Is there an easy way to identify other functions (present or future) that may need similar treatment in our tests?
2. What is it about gebrd that makes the output less directly comparable against LAPACK?

I think some of this was mentioned in discussions a while back, but I don't remember the answers.

cgmb commented on rocsolver/clients/include/testing_gebd2_gebrd.hpp:
```diff
@@ -165,7 +165,7 @@ void gebd2_gebrd_initData(const rocblas_handle handle,
 
 
-template <bool STRIDED, bool GEBRD, typename S, typename T, typename Sd, typename Td, typename Ud, typename Sh, typename Th, typename Uh>
+template <bool STRIDED, bool GEBRD, typename S, typename T, typename Sd, typename Td, typename Ud, typename Sh, typename Th, typename Uh, bool COMPLEX = is_complex<T>>
```
Unless there's a need to override that default value and set `COMPLEX` to something other than the result of `is_complex<T>`, I would just make it a `constexpr` variable within the function.

Type-dependent values are usually simpler when defined via constexpr, with the exception being if you need to sometimes be able to specify them independently from the type.

tfalders replied:
Sure, I can take care of that. The template parameter list is long enough even without the COMPLEX bool.

tfalders replied:
I think I will also turn verify_implicit_test into a constexpr, since it should not change during run time.

cgmb replied:
Sure. There are gotchas for using it on functions and class members, but you can pretty much never go wrong adding constexpr to a local variable.

I would note that I wouldn't expect the compiled code to change in either of these cases, though. The if would be optimized away anyway due to [sparse conditional constant propagation (sccp)](https://en.m.wikipedia.org/wiki/Sparse_conditional_constant_propagation), which clang enables at -O1.

cgmb replied:
I should clarify: I encourage you to use constexpr liberally for local variables. It's a great practice.

The optimizer can often do it's job without the annotation, but the most reliable way to ensure compilers (and other programmers reading your code) know that a variable is a compile-time constant is by marking it as such.

I just mentioned the optimizer behaviour because I think it's worth knowing. Not to dissuade you from your plan.

tfalders commented:
> Not a comment about the code, but maybe I can learn something while this work is fresh in your mind.
> 
> 1. Is there an easy way to identify other functions (present or future) that may need similar treatment in our tests?
> 2. What is it about gebrd that makes the output less directly comparable against LAPACK?
> 
> I think some of this was mentioned in discussions a while back, but I don't remember the answers.

I think the only functions that we need to look out for are those that do not have a unique answer. More specifically, I would suggest being mindful of larfg - to my knowledge, it does not have a unique answer, and the sign of the answer it does give is based on the sign of a potentially small number (thus prone to instability). The factorization computed by gebrd, I believe, is not unique, and the function calls larfg while seemingly providing ample opportunity for larfg to change sign.

jzuniga-amd commented on rocsolver/clients/include/testing_gebd2_gebrd.hpp:
```diff
@@ -196,22 +196,104 @@ void gebd2_gebrd_getError(const rocblas_handle handle,
    
     // error is ||hA - hARes|| / ||hA||
     // (THIS DOES NOT ACCOUNT FOR NUMERICAL REPRODUCIBILITY ISSUES. 
     // IT MIGHT BE REVISITED IN THE FUTURE)
```
You can remove this comment now as we are not comparing with LAPACK anymore.

tfalders commented:
I've addressed the comments and reformatted the code. I'll wait for CI before merging. If there are any other comments, please let me know before then.
